File: src/main/java/net/dv8tion/jda/api/utils/markdown/DiscordDecoration.java
package net.dv8tion.jda.api.utils.markdown;

public enum DiscordDecoration implements IDecoration {
    ITALIC {
        @Override
        public String apply(String text) {
            return IDecoration.around(text, "_");
        }
    },
    BOLD {
        @Override
        public String apply(String text) {
            return IDecoration.around(text, "**");
        }
    },
    UNDERLINE {
        @Override
        public String apply(String text) {
            return IDecoration.around(text, "__");
        }
    },
    STRIKETHROUGH {
        @Override
        public String apply(String text) {
            return IDecoration.around(text, "~~");
        }
    }
}

File: src/main/kotlin/net/gloryx/kda/markdown/DiscordDecoration.kt
package net.gloryx.kda.markdown

enum class DiscordDecoration : Decoration {
    ITALIC {
        override fun apply(text: String): String =
            text.around("_")
    },
    BOLD {
        override fun apply(text: String): String =
            text.around("**")
    },
    UNDERLINE {
        override fun apply(text: String): String =
            text.around("__")
    },
    STRIKETHROUGH {
        override fun apply(text: String): String =
            text.around("~~")
    },
    SINGLE_BLOCK_QUOTE {
        override fun apply(text: String): String = text.split('\n').reduce { acc, it -> "$acc\n> $it"}
    },
    ALL_BLOCK_QUOTE {
        override fun apply(text: String): String = ">>> $text"
    },
    SINGLE_CODE_BLOCK {
        override fun apply(text: String): String = text.split('\n').reduce { acc, it -> "$acc\n`$it`" }
    },
    MULTI_CODE_BLOCK {
        override fun apply(text: String): String = "```\n$text\n```"
    }
}

fun String.around(decoration: String) = "$decoration$this$decoration"